merlin/systems/dag/dictarray.py

```diff
@@ -13,44 +13,149 @@
 #
+from enum import Enum
 from typing import Dict, Optional
 
 import numpy as np
 
-from merlin.core.protocols import SeriesLike, Transformable
+from merlin.core.dispatch import HAS_GPU
+from merlin.core.protocols import SeriesLike
+
+if HAS_GPU:
+    import cupy
+else:
+    cupy = None
+
+
+class Device(Enum):
+    CPU = 0
+    GPU = 1
 
 
 class Column(SeriesLike):
     """
-    A simple wrapper around an array of values
+    A simple wrapper around an array of values. This has an API that's just similar enough to
+    Pandas and cuDF series to be relatively interchangeable from the perspective of the Merlin DAG,
+    but no more. As more methods get added to this class, it gets closer and closer to actually
+    *being* a Pandas/cuDF Series (at which point there's no advantage to using this.) So: keep
+    this class as small as possible.
     """
 
-    def __init__(self, values):
+    def __init__(self, values, row_lengths=None):
         super().__init__()
 
         if isinstance(values, Column):
             raise TypeError("doubly nested columns")
 
         self.values = values
+        self.row_lengths = row_lengths
         self.dtype = values.dtype
 
+        if isinstance(values, np.ndarray):
+            self._device = Device.CPU
+        elif HAS_GPU and cupy and isinstance(values, cupy.ndarray):
+            self._device = Device.GPU
+        else:
+            raise TypeError("type of values argument is not supported")
+
+    def cpu(self):
+        """
+        Move the data for this column to host (CPU) memory
+
+        Returns
+        -------
+        Column
+            Same column, same data but now definitely in CPU memory
+        """
+        self.device = Device.CPU
+        return self
+
+    def gpu(self):
+        """
+        Move the data for this column to device (GPU) memory
+
+        Returns
+        -------
+        Column
+            Same column, same data but now definitely in GPU memory
+        """
+        self.device = Device.GPU
+        return self
+
+    @property
+    def device(self):
+        return self._device
+
+    @device.setter
+    def device(self, device):
+        if device == "cpu":
+            device = Device.CPU
+        elif device == "gpu":
+            device = Device.GPU
+
+        # CPU to GPU
+        if self._device == Device.CPU and device == Device.GPU:
+            self._device_move(cupy.asarray)
+        # GPU to CPU
+        elif self._device == Device.GPU and device == Device.CPU:
+            self._device_move(cupy.asnumpy)
+        # Nothing to do
+        else:
+            return self
+
+    def _device_move(self, fn):
+        if not cupy:
+            raise ValueError("Unable to move data across devices without a GPU")
+        self.values = fn(self.values)
+        if self.row_lengths:
+            self.row_lengths = fn(self.row_lengths)
+
     def __getitem__(self, index):
-        return self.values[index]
+        if self.is_list:
```

Comment:
We're debating whether it's worth splitting list and non-list columns into two separate classes with a shared base class, but it seemed like more trouble than it's worth at this point. I don't love the `if/else` in all these methods though.